crypto/ml_dsa/ml_dsa_sign.c

```diff
@@ -0,0 +1,394 @@
+     */
+    for (kappa = 0; ; kappa += l) {
+        VECTOR *y_ntt = &cs1;
+        VECTOR *r0 = &w1;
+        VECTOR *ct0 = &w1;
+        uint32_t z_max, r0_max, ct0_max, h_ones;
+
+        vector_expand_mask(&y, rho_prime, sizeof(rho_prime), kappa,
+                           gamma1, ctx->h_ctx);
+        vector_copy(y_ntt, &y);
+        vector_ntt(y_ntt);
+
+        matrix_mult_vector(&a_ntt, y_ntt, &w);
+        vector_ntt_inverse(&w);
+
+        vector_high_bits(&w, gamma2, &w1);
+        ossl_ml_dsa_w1_encode(&w1, gamma2, w1_encoded, w1_encoded_len);
+
+        if (!shake_xof_2(h_ctx, mu, sizeof(mu), w1_encoded, 128 * k,
+                         c_tilde, c_tilde_len))
+            break;
+
+        if (!poly_sample_in_ball_ntt(c_ntt, c_tilde, c_tilde_len, ctx->h_ctx,
+                                     params->tau))
+            break;
+
+        vector_mult_scalar(&s1_ntt, c_ntt, &cs1);
+        vector_ntt_inverse(&cs1);
+        vector_mult_scalar(&s2_ntt, c_ntt, &cs2);
+        vector_ntt_inverse(&cs2);
+
+        vector_add(&y, &cs1, &sig.z);
+
+        /* r0 = lowbits(w - cs2) */
+        vector_sub(&w, &cs2, r0);
+        vector_low_bits(r0, gamma2, r0);
+
+        /*
+         * Leaking that the signature is rejected is fine as the next attempt at a
+         * signature will be (indistinguishable from) independent of this one.
+         */
+        z_max = vector_max(&sig.z);
+        r0_max = vector_max_signed(r0);
+        if (value_barrier_32(constant_time_ge(z_max, gamma1 - params->beta)
+                             | constant_time_ge(r0_max, gamma2 - params->beta)))
+            continue;
+
+        vector_mult_scalar(&t0_ntt, c_ntt, ct0);
+        vector_ntt_inverse(ct0);
+        vector_make_hint(ct0, &cs2, &w, gamma2, &sig.hint);
+
+        ct0_max = vector_max(ct0);
+        h_ones = vector_count_ones(&sig.hint);
+        /* Same reasoning applies to the leak as above */
+        if (value_barrier_32(constant_time_ge(ct0_max, gamma2)
+                             | constant_time_lt(params->omega, h_ones)))
+            continue;
+        ret = ossl_ml_dsa_sig_encode(&sig, params, out_sig);
+        break;
+    }
+err:
+    OPENSSL_clear_free(alloc, alloc_len);
+    OPENSSL_cleanse(rho_prime, sizeof(rho_prime));
+    return ret;
+}
+
+/*
+ * See FIPS 204, Algorithm 8, ML-DSA.Verify_internal().
+ */
+static int ml_dsa_verify_internal(ML_DSA_CTX *ctx, const ML_DSA_KEY *pub,
+                                  const uint8_t *msg_enc, size_t msg_enc_len,
+                                  const uint8_t *sig_enc, size_t sig_enc_len)
+{
+    int ret = 0;
+    uint8_t *alloc = NULL, *w1_encoded;
+    POLY *polys = NULL, *p, *c_ntt;
+    MATRIX a_ntt;
+    VECTOR az_ntt, ct1_ntt, *z_ntt, *w1, *w_approx;
+    ML_DSA_SIG sig;
+    const ML_DSA_PARAMS *params = ctx->params;
+    uint32_t k = pub->params->k;
+    uint32_t l = pub->params->l;
+    size_t w1_encoded_len = 128 * k;
```

Comment:
The length `w1_encoded_len` depends on gamma2, and is actually `192 * k` in `ML-DSA-44`.